riscv_isac/coverage.py

```diff
@@ -1006,23 +1005,52 @@ def get_mem_val(addr, bytes):
                     else:
                         return None
 
-            def get_pte_per(pa, pte_addr, pgtb_addr):
-                for match in matches_for_options:
-                    if match[0] == 'VM' and match[1] in ['SV39', 'SV48', 'SV57']:
-                        pte_size = 8
-                    elif match[0] == 'VM' and match[1] in ['SV32']:
-                        pte_size = 4
+            def get_pte(pa, pte_addr, pgtb_addr):
+                if instr_vars['xlen'] == 64:
+                    pte_size = 8
+                elif instr_vars['xlen'] == 32:
+                    pte_size = 4
                 if (pgtb_addr >> 12) == (pte_addr >> 12):
                     if ((pa >> 12) == (get_mem_val(pte_addr, pte_size) >> 10)):
-                        return get_mem_val(pte_addr, pte_size) & 0x3FF
+                        return get_mem_val(pte_addr, pte_size)
                     else:
                         return None
                 else:
                     return None
 
+            def get_pte_prop(prop_name,pa, pte_addr, pgtb_addr):
+                pte_per = get_pte(pa, pte_addr, pgtb_addr)
+                if pte_per is not None:
+                    pte_per = get_pte(pa, pte_addr, pgtb_addr) & 0x3FF
+                    prop_name_lower = prop_name.lower()
+                    if prop_name_lower == 'v' and (pte_per & 0x01 != 0):
+                        return 1
+                    elif prop_name_lower == 'r' and (pte_per & 0x02 != 0):
+                        return 1
+                    elif prop_name_lower == 'w' and (pte_per & 0x04 != 0):
+                        return 1
+                    elif prop_name_lower == 'x' and (pte_per & 0x08 != 0):
+                        return 1
+                    elif prop_name_lower == 'u' and (pte_per & 0x10 != 0):
+                        return 1
+                    elif prop_name_lower == 'g' and (pte_per & 0x20 != 0):
+                        return 1
+                    elif prop_name_lower == 'a' and (pte_per & 0x40 != 0):
+                        return 1
+                    elif prop_name_lower == 'd' and (pte_per & 0x80 != 0):
+                        return 1
+                    elif prop_name_lower in {'rsw', 'RSW'} and (pte_per & 0x300 in {0, 1, 2, 3}):
```

Comment:
This line doesn't look correct. IF you're checking to see if the result is in range {0,1,2,3} then you better have a mask in that range (0x3 - not 0x300) or you should make the range 0x000, 0x100, 0x200, 0x300).
And, regardless of that, if you're checking for 2 bits with 4 possible ranges, it will always match, regardless of the bit positions, because you're checking all possible 2 bit values.  Value 0 is whatyou're looking for, in which case you should check that it is !=0 instead of in {0,1,2,3}